### ISSUE TITLE
🔥 [🐛] Firebase Messaging refuses to work on iOS - Error in application:didFailToRegisterForRemoteNotificationsWithError: (null)

### ISSUE THREAD

**Issue by filiphosko** [invertase/react-native-firebase#8081]
## Issue

Hello, as the title says, I have trouble running Firebase Messaging on iOS. To make it more frustrating, the error messages I get tell me exactly nothing - but maybe you'll be able to decipher them.

I test this in development, with my iPhone 11 connected to my MB Pro (M2 Pro) with Apple's Lightning -> USB-C cable (but I also tried this with a simulator, iPhone 16). I went over the setup multiple times - I have my APN key uploaded into Firebase Console, got the Google plist in my project. I have the Push Notifications and Background Modes capability enabled (with sub modes too). I have Push Notifications enabled in my app identifier. I also modified AppDelegate.mm and the Podfile, run pod install. I also use this library for a long time, so I already did the setup in the past (and it worked), I just needed to upgrade React Native to a more up-to-date version, so I also updated some of the libs. I never had such an issue with this lib where I even don't get a more detailed error message.

I get the first error when the app starts and the Firebase lib starts to initialize modules. Crashes and analytics are enabled fine, with messages I get this error - `[FirebaseMessaging][I-FCM012002] Error in application:didFailToRegisterForRemoteNotificationsWithError: (null)`.

Then in JavaScript I try to request permissions, I get this error - `[messaging/unknown] The operation couldn’t be completed. (UNErrorDomain error 1.)`

---

### Javascript

<details><summary>Click To Expand</summary>
<p>

#### `package.json`:

```json
{
  "name": "MyTroubledApp",
  "version": "0.0.1",
  "private": true,
  "scripts": {
    "android": "react-native run-android",
    "ios": "react-native run-ios",
    "lint": "eslint .",
    "start": "react-native start",
    "test": "jest"
  },
  "dependencies": {
    "@react-native-async-storage/async-storage": "^2.0.0",
    "@react-native-clipboard/clipboard": "^1.14.1",
    "@react-native-firebase/analytics": "^21.2.0",
    "@react-native-firebase/app": "^21.2.0",
    "@react-native-firebase/crashlytics": "^21.2.0",
    "@react-native-firebase/messaging": "^21.2.0",
    "@react-native-masked-view/masked-view": "^0.3.1",
    "@react-native-picker/picker": "^2.8.1",
    "@react-navigation/drawer": "^6.7.2",
    "@react-navigation/elements": "^1.3.31",
    "@react-navigation/native": "^6.1.18",
    "@react-navigation/stack": "^6.4.1",
    "@reduxjs/toolkit": "^2.2.7",
    "ably": "^2.3.1",
    "color": "^4.2.3",
    "js-base64": "^3.7.7",
    "lodash-es": "^4.17.21",
    "moment": "^2.30.1",
    "native-base": "^2.15.2",
    "react": "18.3.1",
    "react-native": "0.75.2",
    "react-native-calendars": "^1.1306.0",
    "react-native-exit-app": "^2.0.0",
    "react-native-extended-stylesheet": "^0.12.0",
    "react-native-gesture-handler": "^2.18.1",
    "react-native-linear-gradient": "^2.8.3",
    "react-native-reanimated": "^3.15.1",
    "react-native-restart": "^0.0.27",
    "react-native-safe-area-context": "^4.10.9",
    "react-native-screens": "^3.34.0",
    "react-native-swiper": "^1.6.0",
    "react-redux": "^9.1.2",
    "redux-persist": "^6.0.0",
    "shortid": "^2.2.16"
  },
  "devDependencies": {
    "@babel/core": "^7.20.0",
    "@babel/preset-env": "^7.20.0",
    "@babel/runtime": "^7.20.0",
    "@react-native/babel-preset": "0.75.2",
    "@react-native/eslint-config": "0.75.2",
    "@react-native/metro-config": "0.75.2",
    "@react-native/typescript-config": "0.75.2",
    "@types/lodash-es": "^4.17.12",
    "@types/react": "^18.2.6",
    "@types/react-test-renderer": "^18.0.0",
    "babel-jest": "^29.6.3",
    "eslint": "^8.19.0",
    "jest": "^29.6.3",
    "prettier": "2.8.8",
    "react-native-dotenv": "^3.4.11",
    "react-test-renderer": "18.3.1",
    "typescript": "5.0.4"
  },
  "resolutions": {
    "@react-native-picker/picker": "2.8.1"
  },
  "engines": {
    "node": ">=18"
  },
  "packageManager": "yarn@3.6.4"
}
```

#### `firebase.json` for react-native-firebase v6:

```json
{
  "react-native": {
    "crashlytics_debug_enabled": false
  }
}
```

</details>

### iOS

<details><summary>Click To Expand</summary>
<p>

#### `ios/Podfile`:

- [ ] I'm not using Pods
- [x] I'm using Pods and my Podfile looks like:

```ruby
# Resolve react_native_pods.rb with node to allow for hoisting
require Pod::Executable.execute_command('node', ['-p',
  'require.resolve(
    "react-native/scripts/react_native_pods.rb",
    {paths: [process.argv[1]]},
  )', __dir__]).strip

platform :ios, min_ios_version_supported
prepare_react_native_project!

$RNFirebaseAnalyticsWithoutAdIdSupport = true

linkage = ENV['USE_FRAMEWORKS']
if linkage != nil
  Pod::UI.puts "Configuring Pod with #{linkage}ally linked Frameworks".green
  use_frameworks! :linkage => linkage.to_sym
end

target 'MyTroubledApp' do
  config = use_native_modules!
  
  use_frameworks! :linkage => :static
  $RNFirebaseAsStaticFramework = true

  use_react_native!(
    :path => config[:reactNativePath],
    # An absolute path to your application root.
    :app_path => "#{Pod::Config.instance.installation_root}/.."
  )

  target 'MyTroubledAppTests' do
    inherit! :complete
    # Pods for testing
  end

  post_install do |installer|
    # https://github.com/facebook/react-native/blob/main/packages/react-native/scripts/react_native_pods.rb#L197-L202
    react_native_post_install(
      installer,
      config[:reactNativePath],
      :mac_catalyst_enabled => false,
      # :ccache_enabled => true
    )
  end
end
```

#### `AppDelegate.m`:

```objc
#import "AppDelegate.h"
#import <Firebase.h>

#import <React/RCTBundleURLProvider.h>

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
  [FIRApp configure];
  self.moduleName = @"MyTroubledApp";
  // You can add your custom initial props in the dictionary below.
  // They will be passed down to the ViewController used by React Native.
  self.initialProps = @{};

  return [super application:application didFinishLaunchingWithOptions:launchOptions];
}

- (NSURL *)sourceURLForBridge:(RCTBridge *)bridge
{
  return [self bundleURL];
}

- (NSURL *)bundleURL
{
#if DEBUG
  return [[RCTBundleURLProvider sharedSettings] jsBundleURLForBundleRoot:@"index"];
#else
  return [[NSBundle mainBundle] URLForResource:@"main" withExtension:@"jsbundle"];
#endif
}

@end
```

</p>
</details>

---

### Android

<details><summary>Click To Expand</summary>
<p>

#### Have you converted to AndroidX?

- [ ] my application is an AndroidX application?
- [ ] I am using `android/gradle.settings` `jetifier=true` for Android compatibility?
- [ ] I am using the NPM package `jetifier` for react-native compatibility?

#### `android/build.gradle`:

```groovy
// N/A
```

#### `android/app/build.gradle`:

```groovy
// N/A
```

#### `android/settings.gradle`:

```groovy
// N/A
```

#### `MainApplication.java`:

```java
// N/A
```

#### `AndroidManifest.xml`:

```xml
<!-- N/A -->
```

</p>
</details>

---

## Environment

<details><summary>Click To Expand</summary>
<p>

**`react-native info` output:**

```
System:
  OS: macOS 14.7
  CPU: (12) arm64 Apple M2 Pro
  Memory: 386.55 MB / 16.00 GB
  Shell:
    version: "5.9"
    path: /bin/zsh
Binaries:
  Node:
    version: 21.7.1
    path: ~/.asdf/installs/nodejs/21.7.1/bin/node
  Yarn:
    version: 3.6.4
    path: ~/.asdf/installs/nodejs/21.7.1/bin/yarn
  npm:
    version: 10.5.0
    path: ~/.asdf/plugins/nodejs/shims/npm
  Watchman:
    version: 2024.08.19.00
    path: /opt/homebrew/bin/watchman
Managers:
  CocoaPods:
    version: 1.15.2
    path: /Users/filiphosko/.asdf/shims/pod
SDKs:
  iOS SDK:
    Platforms:
      - DriverKit 24.0
      - iOS 18.0
      - macOS 15.0
      - tvOS 18.0
      - visionOS 2.0
      - watchOS 11.0
  Android SDK: Not Found
IDEs:
  Android Studio: 2024.1 AI-241.18034.62.2411.12169540
  Xcode:
    version: 16.0/16A242d
    path: /usr/bin/xcodebuild
Languages:
  Java:
    version: 22.0.2
    path: /usr/bin/javac
  Ruby:
    version: 3.3.0
    path: /Users/filiphosko/.asdf/shims/ruby
npmPackages:
  "@react-native-community/cli": Not Found
  react:
    installed: 18.3.1
    wanted: 18.3.1
  react-native:
    installed: 0.75.2
    wanted: 0.75.2
  react-native-macos: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: Not found
  newArchEnabled: Not found
iOS:
  hermesEnabled: Not found
  newArchEnabled: false
```

- **Platform that you're experiencing the issue on**:
  - [x] iOS
  - [ ] Android
  - [ ] **iOS** but have not tested behavior on Android
  - [ ] **Android** but have not tested behavior on iOS
  - [ ] Both
- **`react-native-firebase` version you're using that has this issue:**
  - 21.2.0, but I had this issue with an older version too, I just upgraded in the midst of debugging
- **`Firebase` module(s) you're using that has the issue:**
  - `e.g. Instance ID`
- **Are you using `TypeScript`?**
  - `Y/N` & `VERSION`

</p>
</details>

---

- 👉 Check out [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter for updates on the library.

**Comment by mikehardy** [comment]
I saw that :

> Then in JavaScript I try to request permissions, I get this error - [messaging/unknown] The operation couldn’t be completed. (UNErrorDomain error 1.)


As well when I was trying to allow messaging permission, when I uninstalled the app, reinstalled it, and restarted the simulator - some combination of that allowed the simulator to finally have things working.

I recognize that's vague, and I apologize for that but I'm pretty methodical when I'm testing things on Notifee and react-native-firebase and I swear I saw that and either a restart or an uninstall/reinstall cycle allowed me to finally get the permissions prompt, accept it, and then things worked for me ? Perhaps worth a try

**Comment by filiphosko** [comment]
@mikehardy Thanks a lot Mike, I really appreciate you reply - I thought that maybe it could be a glitch in development, I already spent several hours debugging it. I will try doing some reinstalling/restarting, although I already tried that on my iPhone (physical device). What I want to try is to build an ad-hoc staging build to check if the same thing will happen there, just to be sure it's not a dev environment type of glitch.

**Comment by filiphosko** [comment]
@mikehardy So, after playing with it for couple of hours I still can't make it work. Tried it with multiple simulators (M2 MB Pro) and my iPhone 11. Cleaned the project multiple times, reinstalled, nothing helped.

What I tried now most recently was to disable auto init in firebase.json and do the registration manually in the app. I stumbled upon some weird behaviour, that I don't completely understand. The steps are as follows:

1) With auto init disabled in firebase.json, I run `messaging().registerDeviceForRemoteMessages()` early in my app
2) I get two errors, `11.4.0 - [FirebaseMessaging][I-FCM012002] Error in application:didFailToRegisterForRemoteNotificationsWithError: (null)` and `[Error: [messaging/unknown] The operation couldn’t be completed. (RNFBErrorDomain error 666.)]`
3) I then proceed to the app where I run `messaging().isAutoInitEnabled`, I get `true`, and `messaging().isDeviceRegisteredForRemoteMessages`, I get `true` also.
4) I proceed further, I run `messaging() .requestPermission()`, I get an error `[Error: [messaging/unknown] The operation couldn’t be completed. (UNErrorDomain error 1.)]`
5) Then I get `-[RNFBMessagingModule registerForRemoteNotifications::]_block_invoke_2 [Line 335] RNFBMessaging dispatch_after: registerDeviceForRemoteMessages handled.`

How is it possible that I get `isAutoInitEnabled` to be `true` if I specifically disabled it with `"messaging_ios_auto_register_for_remote_messages": false` and also `isDeviceRegisteredForRemoteMessages` to be `true` if it failed?

**Comment by filiphosko** [comment]
@mikehardy No problem, I really appreciate all your efforts. Maybe hold on a bit with the repro - I want to try to create a fresh RN project and try it all again. It's possible that something got messed up after I upgraded RN - I usually start with a fresh project and just copy/paste my files there, since I didn't have the best experience upgrading RN in the past, this is how I did it this time too - but maybe something just went wrong there, I don't know.

I was just curious about the errors I was getting since they don't make much sense to me.

**Comment by mikehardy** [comment]
Perhaps if new project then on iOS the background modes and notification capabilities aren't configured? It'll be a while before I have a good reproduction scenario regardless, even though I need it for my next batch of Notifee work...

**Comment by filiphosko** [comment]
I created a brand new project now and it works, so probably something got messed up, I don't know what. I had capabilities setup correctly in the other one that gave me errors too - it's exactly the same setup I have now, the only difference was that I setup different configurations/schemas for debug, staging, etc. that I don't have yet in this new one.

This is a vanilla project, so now I'll start migrating my TS code there and adding other 3rd party libs (also native, with pods) that we use in the project, I'll let you know in case it breaks in the process - I'll try to go incrementally to make sure I know what broke it if it breaks.

**Comment by mikehardy** [comment]
Okay - good to hear you've got some positive forward progress
I'm going to close this as I don't believe it's actionable in this repo but it's just one button click away from reopening if you discover something in here and we need to look. Good luck with your project

**Comment by savioPimenta** [comment]
I had the same error. [Searching on the Apple forum](https://forums.developer.apple.com/forums/thread/660648?answerId=634118022#634118022), I discovered that the error "UNErrorDomain error 1" has to do with the app's Product Name. My Product Name had a special character "ˆ" and this prevented the modal that requests notification permission from being displayed.

1 - I removed the special character through XCode in Targets > MyProject > BuildSettings > Packaging > Product Name
2 - In the Info.plist, I changed the "CFBundleDisplayName" field from "$(PRODUCT_NAME)" to the name I wanted to give my app (with the special character)

As you can see, the special character was kept in the notification popup and in the app's displayed name, but now it works.

![Image](https://github.com/user-attachments/assets/424f579d-bf91-441c-b0bb-d3ceeb67b4d9)

**Comment by filiphosko** [comment]
@savioPimenta This was probably the issue for my project too since it also has a special character in its name. I "fixed" it without knowing by just init-ing a fresh RN project where the name wasn't setup yet and then I just did the Product Name differently, I created a custom field in build settings and used that in my Info.plist (since I have different app configurations/schemes with different name suffixes, like "staging", etc.). But good to know stuff like this can break push notifications, maybe this post will help somebody in the future.
